### PR TITLE
Add support for CLI

### PR DESCRIPTION
**Pull request by thedrummeraki**
#17

### Changes

⚠️ This PR is ready for review, but is missing some tests, which will be added at a later time.

- Dependencies added:
  - `clap`: For accepting CLI parameters and arguments
- Refactoring:
  - Create a `lib.rs` to better organize the overall structure of the code.
  - Create `cli` module: houses all CLI-related code. May grow in the future to accept more arguments.
  - Create `domain` module: contains all domain-specific business logic (types, errors, config, etc.)
- New changes:
  - Create `WorkersList` strongly typed structure.
    - Why? It's tricky to assert for empty vectors using `clap`, so we're taking advantage of the `From` trait to convert the params into a `Config`, which will, when parsed, will be responsible if that there's at least 1 worker to target.
  - Create `LoadBalancerError` enum and `LoadBalancerResult<T>` type to catch all errors in the app.

### Checklist

- ❌ I haven't written tests for my code 
- [x] It is safe to simply rollback this change
- [x] This change does NOT contain a DB migration/rollback

### Testing evidence

- [x] I have provided some testing evidence below (in the form of: deployment link, screenshot, etc...)

👇 Load Balancer running
image

👇 Testing against the Load Balancer
> What is this `ping-tanoshimu`? This is a server I'm working on, on the side. Port `3000` runs the GraphQL API, Port `3001` runs a proxy to this server. Different headers are returned, which is how I can confirm that everything is working properly. The command is an alias to `curl`.

image

### CODE REVIEW

File: src/domain/error.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+pub type LoadBalancerResult<T> = Result<T, LoadBalancerError>;
+
+#[derive(Debug)]
```
Comment (luiscarlosjayk) [inline src/domain/error.rs:3]: I'd suggest considering usage of `thiserror` crate or similar.

File: src/domain/load_balancer.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+pub async fn run(config: Config) -> LoadBalancerResult<()> {
+    let worker_hosts = config.worker_hosts.to_owned();
+    println!("Serving loading balancer on:");
+    println!("{worker_hosts}");
+
+    let load_balancer = Arc::new(RwLock::new(
+        LoadBalancer::new(worker_hosts.as_ref().to_vec()).expect("failed to create load balancer"),
+    ));
+
+    let addr: SocketAddr = SocketAddr::from(([127, 0, 0, 1], 1337));
```
Comment (luiscarlosjayk) [inline src/domain/load_balancer.rs:84]: I think hardcoded port `1337` should come from the `config` parameter.

File: src/domain/load_balancer.rs
Comment (luiscarlosjayk) [inline src/domain/load_balancer.rs:1]: I wonder if there might be any difference to defer the Write Lock just to update the current worker index.

Like this:

```rs
impl LoadBalancer {
    pub fn new(worker_hosts: Vec<String>) -> Result<Self, String> {
        if worker_hosts.is_empty() {
            return Err("No worker hosts provided".into());
        }

        Ok(LoadBalancer {
            client: Client::new(),
            worker_hosts,
            current_worker: 0,
        })
    }

    pub fn forward_request(&self, req: Request<Body>) -> ResponseFuture {
        let mut worker_uri = self.get_worker().to_owned();

        // Extract the path and query from the original request
        if let Some(path_and_query) = req.uri().path_and_query() {
            worker_uri.push_str(path_and_query.as_str());
        }

        // Create a new URI from the worker URI
        let new_uri = Uri::from_str(&worker_uri).unwrap();

        // Extract the headers from the original request
        let headers = req.headers().clone();

        // Clone the original request's headers and method
        let mut new_req = Request::builder()
            .method(req.method())
            .uri(new_uri)
            .body(req.into_body())
            .expect("request builder");

        // Copy headers from the original request
        for (key, value) in headers.iter() {
            new_req.headers_mut().insert(key, value.clone());
        }

        self.client.request(new_req)
    }

    fn get_worker(&self) -> &str {
        // Use a round-robin strategy to select a worker
        let worker = self.worker_hosts.get(self.current_worker).unwrap();
        // self.current_worker = (self.current_worker + 1) % self.worker_hosts.len();
        worker
    }
}

async fn handle(
    req: Request<Body>,
    load_balancer: Arc<RwLock<LoadBalancer>>,
) -> Result<Response<Body>, hyper::Error> {
    let load_balancer_reader = load_balancer.read().await;
    let response = load_balancer_reader.forward_request(req).await;

    // Delay updating current_worker index 
    let mut load_balancer = load_balancer.write().await;
    let next_current_worker = load_balancer
        .current_worker
        .rem_euclid(load_balancer.worker_hosts.len());
    load_balancer.current_worker = next_current_worker;

    response
}
```

File: src/domain/worker.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+impl WorkersList {
+    pub fn parse(list: Vec<String>) -> LoadBalancerResult<Self> {
+        if list.is_empty() {
+            return Err(LoadBalancerError::ConfigError(
+                "At least 1 worker must be specified.".into(),
+            ));
+        }
+
+        Ok(Self(list))
+    }
```
Comment (letsgetrusty) [inline src/domain/worker.rs:17]: Great use of the "parse don't validate" pattern! 🔥 
You could additionally validate that the strings themselves are properly formatted.

File: src/domain/worker.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+impl AsRef<Vec<String>> for WorkersList {
+    fn as_ref(&self) -> &Vec<String> {
+        &self.0
+    }
+}
+
+impl Display for WorkersList {
```
Comment (letsgetrusty) [inline src/domain/worker.rs:26]: Nice use-case for the Display trait!

**Review by letsgetrusty — APPROVED** [review]
Overall LGTM!

File: src/cli/mod.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+struct Serve {
+    /// Specify the port to run the server on.
+    #[clap(short, long, default_value = "1337")]
+    port: u16,
+
+    #[clap(short, long)]
+    worker: Vec<String>,
+}
+
+impl TryFrom<Serve> for Config {
```
Comment (cj-zhukov) [inline src/cli/mod.rs:36]: Look's like a convenient way to get Config! I haven't thought about this approach.

File: src/domain/error.rs (hunk 1 above)
Comment (thedrummeraki) [reply src/domain/error.rs:3]: Yes, good point. I've created an issue to address this: https://github.com/thedrummeraki/capstone-rust/issues/21

File: src/domain/load_balancer.rs
Comment (thedrummeraki) [reply src/domain/load_balancer.rs:1]: Very good points here as well. Will definitely address those in an upcoming PR, as this code was mostly adapted from the Load Balancer template. I'll also create an issue for this.

**Pull request merged**